### PR TITLE
CI: Adding script to validate consistent and correct capitalization among headings in documentation (#26941)

### LINKED ISSUE #26941: DOC: Validate consistency of title capitalization
In #26933, we're making the capitalization of the title sections consisten. We use to have many titles capitalized as `This is the Section Title`, and we changed all them (probably few were forgotten) to `This is the section title`.

To keep this consistency, we should validate that the capitalization is correct in the CI. This can be done by extracting all the titles, and making sure that only the first letter of the sentence is uppercase, or words defined in a short list, like `Series`, `DataFrame`,...

I think this can be done in two ways:
- As a sphinx extension that validates the titles as they are processed, and generates warnings if they are not (this will automatically fail the CI).
- As an independent script

The first option should be simpler if sphinx can implement this as extension, but not sure if that's the case.

### PR DESCRIPTION
**Pull request by tonywu1999**
Closes #26941

This script should be running correctly to validate capitalization for various rst files.  Many keywords will need to be added into the script as more and more files are tested on this script. Currently testing on the following rst files through code_checks.sh:

doc/source/development/contributing.rst
doc/source/index.rst doc/source/ecosystem.rst

contributing.rst and ecosystem.rst should produce an exit code of 1, indicating there are incorrect capitalization in their headings.

### CODE REVIEW

**Comment by WillAyd** [comment]
This is a lot of code; is there no third party package to do something like this?

**Comment by tonywu1999** [comment]
@WillAyd 

I've looked around when trying to solve this problem and I don't think there's a third party package that does everything that we need.  Ultimately, the more complicated parts of the script involve the combination of finding headers in an RST file & determining if a string follows our capitalization convention while ensuring keywords like pandas are properly capitalized too.

I will keep looking for third party packages online to hopefully simplify the lines of this script.  In the meantime, my goal right now is to check if the script is working as expected.

File: ci/code_checks.sh [hunk 1]
```diff
@@ -318,6 +318,11 @@ if [[ -z "$CHECK" || "$CHECK" == "docstrings" ]]; then
     $BASE_DIR/scripts/validate_docstrings.py --format=actions --errors=GL03,GL04,GL05,GL06,GL07,GL09,GL10,SS04,SS05,PR03,PR04,PR05,PR10,EX04,RT01,RT04,RT05,SA02,SA03,SA05
     RET=$(($RET + $?)) ; echo $MSG "DONE"
 
+    ### VALIDATE TITLE CAPITALIZATION CONVENTION ###
```
Comment (datapythonista) [inline ci/code_checks.sh]: I think this comment is redundant with the message in next line, you can probably get rid of it

File: ci/code_checks.sh [hunk 2]
```diff
@@ -340,5 +345,4 @@ if [[ -z "$CHECK" || "$CHECK" == "typing" ]]; then
     RET=$(($RET + $?)) ; echo $MSG "DONE"
 fi
 
-
```
Comment (datapythonista) [inline ci/code_checks.sh]: Try to avoid unrelated changes, even if they're small

File: scripts/validate_rst_title_capitalization.py [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+        for fd in self.null_fds + self.save_fds:
+            os.close(fd)
+
+
+# Keynames that would not follow capitalization convention
+CAPITALIZATION_EXCEPTIONS = {
+    'pandas', 'Python', 'IPython', 'PyTables', 'Excel', 'JSON',
+    'HTML', 'SAS', 'SQL', 'BigQuery', 'STATA', 'Interval', 'PEP8',
+    'Period', 'Series', 'Index', 'DataFrame', 'C', 'Git', 'GitHub', 'NumPy',
+    'Apache', 'Arrow', 'Parquet', 'Triage', 'MultiIndex', 'NumFOCUS', 'sklearn-pandas'
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: I don't think triage should be capitalized, I can be wrong

File: scripts/validate_rst_title_capitalization.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+}
+
+# Lowercase representation of CAPITALIZATION_EXCEPTIONS
+CAPITALIZATION_EXCEPTIONS_LOWER = {word.lower() for word in CAPITALIZATION_EXCEPTIONS}
+
+# Dictionary of bad titles that will be printed later along with line numbers
+# Key: Document Directory, Value: Pair(Bad Title, Line Number)
+badTitleDictionary = {}
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: In Python the convention is to use `bad_title_dictionary` instead of camel case. This applies to all variables, and function names in your code. Have a look at the PEP-8 style guide.

File: scripts/validate_rst_title_capitalization.py [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+    to stderr just before a script exits, and after the context manager has
+    exited (at least, I think that is why it lets exceptions through).
+
+    This code is needed to suppress output from the parser method
+    because the parser method prints to stdout when encountering Sphinx
+    references, as it cannot parse those at this moment.
+
+    '''
+    def __init__(self):
+        # Open a pair of null files
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Better avoid the comments where you explain the code, python code is very readable. Not the init, but for the other functions of you add a comment explaining what's their goal, use a docstring instead of a comment with a hash

File: scripts/validate_rst_title_capitalization.py [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+        elif (directoryAddress.endswith(".rst")):
+            f.append(directoryAddress)
+        else:
+            for (dirpath, dirnames, filenames) in walk(directoryAddress):
+                for file in filenames:
+                    if file.endswith(".rst"):
+                        f.append(os.path.join(dirpath, file))
+
+    # Return the filled up list of .rst file paths
+    return f
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: I think it's probably better to use a generator for this. Also, I'd avoid making the function "create", I think it's misleading (like creating a directory in disk...)

**Review by datapythonista — COMMENTED** [review]
Nice work. You probably want to check the other scripts, you'd see were very consistent with styles, and you'll need to adapt. But looks good, the context manager to supress the output is probably the best option.

Thanks!

**Comment by tonywu1999** [comment]
I'm seeing some problems in the CI/Checks that are occurring outside of the errors I purposely created myself.  I'm not exactly sure if my script is causing these errors or if these are errors independent of my script.

1. Linting
Check import format using isort
##[error]Skipped 2 files
Check import format using isort DONE
##[error]Process completed with exit code 1

2. Publish benchmarks artifacts
##[error]Path does not exist /home/runner/work/pandas/pandas/asv_bench/benchmarks.log
##[error]Exit code 1 returned from process: file name '/home/runner/runners/2.163.1/bin/Runner.PluginHost', arguments 'action "GitHub.Runner.Plugins.Artifact.PublishArtifact, Runner.Plugins"'.

Furthermore, I'm having trouble suppressing stderr output that is produced from an imported helper method.  When I attempt to suppress the stderr output on my local machine using a context manager (_class suppress_stdout_stderr_ on line 35), the suppressing works.  However, only when I push my code onto GitHub does the suppressing fail to work on CI/Checks.  This causes more clutter than I would want to appear on CI/Checks.

**Comment by datapythonista** [comment]
I'm in my phone and can't access github actions, but I think both errors are false positives that fail in all PRs, but don't make the CI fail. Is it possible that your script is not exiting with a zero code, or that another error is happening?

**Comment by tonywu1999** [comment]
Ok, so to reiterate what you said, I do not need to worry about the 'linting' and 'publish benchmark artifacts' checks.  

Also, I set an exit code of 1 in validate_rst_title_capitalization.py if there are titles/headings that have improper capitalization.  Currently, the script is checking to validate a file that has headings with improper capitalization, so it is returning an exit code of 1 at the moment.  Is this the reason why the context manager is able to suppress stderr output on my local machine but not on GitHub?  I'm very confused on what is going on with regards to suppressing output.

**Comment by datapythonista** [comment]
Just ignore those errors then.

I don't think what I said is related to the output not being suppressed in github, not sure about that.

**Comment by MomIsBestFriend** [comment]
@tonywu1999 thank you for the code.

You'll need to run "black" on the new created file.

```bash
$ black  path/to/your/file
```

(Don't panic, but it's going to change a lot of lines).

**Comment by MomIsBestFriend** [comment]
> I'm in my phone and can't access github actions, but I think both errors are false positives that fail in all PRs, but don't make the CI fail. Is it possible that your script is not exiting with a zero code, or that another error is happening?

```Options -> Request Desktop site``` :)

cc @datapythonista

**Comment by tonywu1999** [comment]
@datapythonista 

Because the suppressing wasn't working, I changed my code so that rather than using a third party module to find headings in an RST file, I replaced the module with my own code from scratch.  Now, I'm fairly confident the script is outputting on CI/Checks exactly what I want it to output.  

Where should I go from here?  Also if a file contains headings that are not properly capitalized, should I return an system exit code other than zero?

I also noticed a new error in 'continuous-integration/travis-ci/pr', and I'm not sure if that is related to anything in my script.

File: scripts/validate_rst_title_capitalization.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+
+# Dictionary of bad titles that will be printed later along with line numbers
+# Key: Document Directory, Value: Pair(Bad Title, Line Number)
+bad_title_dict = {}
+
+# Error Message:
+err_msg = "Heading capitalization formatted incorrectly. Please correctly capitalize"
+
+
+def follow_capitalization_convention(title: str) -> bool:
```
Comment (MomIsBestFriend) [inline scripts/validate_rst_title_capitalization.py]: Since this is returning a bool, I think that ```is_following_capitalization_convention``` is more appropriate function name.

File: scripts/validate_rst_title_capitalization.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+    # title_list is the list of headings that is encountered in the doctree
+    title_list: List[str] = []
+
+    # List of line numbers that corresponding headings in title_list can be found at
+    line_number_list: List[int] = []
+
+    # Open and read the .rst file and store the string of data into input
+    f = open(rst_file, "r")
+    input = f.read().split("\n")
+    f.close()
```
Comment (MomIsBestFriend) [inline scripts/validate_rst_title_capitalization.py]: Can you use a contex manager here?

I believe that

```python
with open(rst_file, "r") as file_obj:
    lines = file_obj.read().split("\n")
```

Should do the trick here.

side note, ```input``` is a built-in python function, please try to avoid using it as a variable name

File: scripts/validate_rst_title_capitalization.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    # Append the bad_title_dict if the capitalization convention not followed
+    for i in range(len(title_list)):
+        if not follow_capitalization_convention(title_list[i]):
+            if rst_file not in bad_title_dict:
+                bad_title_dict[rst_file] = [(title_list[i], line_number_list[i])]
+            else:
+                bad_title_dict[rst_file].append((title_list[i], line_number_list[i]))
+
+
+def find_rst_files(source_paths: List[str]) -> List[str]:
```
Comment (MomIsBestFriend) [inline scripts/validate_rst_title_capitalization.py]: I think that the return type here is:

```python
Generator[str, None, None]
```

Comment (tonywu1999) [reply scripts/validate_rst_title_capitalization.py]: @MomIsBestFriend Thank you for your input!  It's really helping me a lot.

File: scripts/validate_rst_title_capitalization.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+}
+
+# Lowercase representation of CAPITALIZATION_EXCEPTIONS
+CAPITALIZATION_EXCEPTIONS_LOWER = {word.lower() for word in CAPITALIZATION_EXCEPTIONS}
+
+# Dictionary of bad titles that will be printed later along with line numbers
+# Key: Document Directory, Value: Pair(Bad Title, Line Number)
+bad_title_dict = {}
+
+# Error Message:
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Please delete

File: scripts/validate_rst_title_capitalization.py [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+    """
+
+    # Remove https link if present in heading
+    title = re.sub(r"<https?:\/\/.*[\r\n]*>", "", title)
+
+    # Split with delimiters comma, semicolon and space, parentheses, colon, slashes
+    word_list = re.split(r"[;,-/():\s]\s*", title)
+
+    # Edge Case: First word is an empty string
+    if len(word_list[0]) == 0:
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: `if not word_list[0]:`

Do you know why is this happening?

File: scripts/validate_rst_title_capitalization.py [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+        return False
+
+    # Dealing with the first word of the title
+    if word_list[0] not in CAPITALIZATION_EXCEPTIONS:
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: I think this should be part of the next loop, inside the loop check first if the word is a proper name, which is common for all words, and then have an if to see if it's the first word or not (you can use `enumerate` in the loop for it)

File: scripts/validate_rst_title_capitalization.py (hunk 9 above)
Comment (MomIsBestFriend) [reply scripts/validate_rst_title_capitalization.py]: No problems, and if I gotta be honest I learnt a lot from @datapythonista, I would suggest you to read his code suggestions he gave me in https://github.com/pandas-dev/pandas/pull/30467 and in https://github.com/pandas-dev/pandas/pull/30755

I think that I stepped in pretty much every possible pitfall there is.

File: scripts/validate_rst_title_capitalization.py (hunk 12 above)
Comment (datapythonista) [reply scripts/validate_rst_title_capitalization.py]: Or better implementation could be to have a function `correct_title_capitlization(title), that simply calls the string function to make the first letter upper case and ask the rest lower case, and then replace proper words of any from the list.

This way you can add in the error message what is the correct capitalization.

File: scripts/validate_rst_title_capitalization.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+        Output format of the script.
+
+    Returns
+    -------
+    is_failed : bool
+        True if there are headings that are printed, False if not
+
+    """
+
+    is_failed: bool = False
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: I would count how many errors are there and return the counter as the exit code

File: scripts/validate_rst_title_capitalization.py [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+            # Remaining letters must not be uppercase
+            for j in range(len(word_list[i])):
+                if word_list[i][j].isupper():
+                    return False
+
+    # Returning True if the heading follows the capitalization convention
+    return True
+
+
+def findTitles(rst_file: str) -> Generator[List[str], List[int], None]:
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Avoid camel case

File: scripts/validate_rst_title_capitalization.py [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+    ----------
+    rst_file : str
+        RST file to scan through for headings
+
+    Returns
+    -------
+    title_list : List[str]
+        A list of heading strings found in the document tree
+
+    line_number_list : List[int]
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Return tuples of title de longe number instead of two lists

**Review by datapythonista — COMMENTED** [review]
Nice work, added some more comments X but this is looking much better. Please avoid comments explaining the code, unless it's not possible that the code is clear enough (which happens very rarely in Python).

But this starts to look much better.

The exit code, as I said in a comment, my preference is returning the number of errors found.

Thanks a lot for working on this!

File: scripts/validate_rst_title_capitalization.py [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+
+    # Return an exit status of 0 if there are no bad titles in the dictionary
+    if len(bad_title_dict) == 0:
+        return is_failed
+
+    # Print bad_title_dict Results
+    is_failed = True
+    for key in bad_title_dict:
+        for line in bad_title_dict[key]:
+            print(key + ":" + str(line[1]) + ": " + err_msg + ' "' + line[0] + '"')
```
Comment (MomIsBestFriend) [inline scripts/validate_rst_title_capitalization.py]:
```python
print(f'{key}:{line[1]}:{err_msg} "{line[0]"')
```

File: scripts/validate_rst_title_capitalization.py [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+    # Loop through source_paths, recursively looking for .rst files
+    for directory_address in source_paths:
+        if not os.path.exists(directory_address):
+            raise ValueError(
+                "Please enter a valid path, pointing to a valid file/directory."
+            )
+        elif directory_address.endswith(".rst"):
+            yield directory_address
+        else:
+            for (dirpath, dirnames, filenames) in walk(directory_address):
```
Comment (MomIsBestFriend) [inline scripts/validate_rst_title_capitalization.py]: ```dirnames``` never gets used, can you replace it with ```_```?

should like something like this:

```python
for dirpath, _, filenames in os.walk(directory_address):
```

File: scripts/validate_rst_title_capitalization.py (hunk 10 above)
Comment (tonywu1999) [reply scripts/validate_rst_title_capitalization.py]: Could you clarify what I should be deleting here?

File: scripts/validate_rst_title_capitalization.py (hunk 11 above)
Comment (tonywu1999) [reply scripts/validate_rst_title_capitalization.py]: I came across situations where the text in a heading was like the following:

'  heading'

where there was some number of spaces in the beginning of the heading.  In these situations, the split function would cause the first word in the word_list array to be an empty string.

Comment (tonywu1999) [reply scripts/validate_rst_title_capitalization.py]: In this situation, should the heading be considered incorrect?

**Comment by tonywu1999** [comment]
I edited the script so that now there is a function (correct_title_capitalization) that let's the user know what the corrected capitalization title should be.

Example output:
ci/../doc/source/development/contributing.rst:824:Heading capitalization formatted incorrectly. Please correctly capitalize "Type Hints" to "Type hints"

File: scripts/validate_rst_title_capitalization.py [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+        True if there are headings that are printed, False if not
+
+    """
+
+    number_of_errors: int = 0
+
+    directory_list = find_rst_files(source_paths)
+
+    for filename in directory_list:
+        fill_bad_title_dict(filename)
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: This F feels overcomplicated. Just loop over `find_titles` here, and if `title == correct_title_capitalization(title)` yield a tuple with the file name, and the correct and incorrect titles.

I think you can get rid a lot of the code by doing that, and things will be much clearer.

File: scripts/validate_rst_title_capitalization.py [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+    correct_title: str = re.sub(r"^\W*", "", title).capitalize()
+
+    removed_https_title = re.sub(r"<https?:\/\/.*[\r\n]*>", "", correct_title)
+
+    word_list = re.split(r"\W", removed_https_title)
+
+    for word in word_list:
+        if word.lower() in CAP_EXCEPTIONS_DICT:
+            correct_title = re.sub(
+                r"\b" + word + r"\b", CAP_EXCEPTIONS_DICT[word.lower()], correct_title
```
Comment (MomIsBestFriend) [inline scripts/validate_rst_title_capitalization.py]: Can you please use f-strings here?

```python
correct_title = re.sub(rf"\b{word}\b", CAP_EXCEPTIONS_DICT[word.lower()], correct_title)
```

**Comment by tonywu1999** [comment]
I edited the script so the main function has been simplified.  I got rid of the bad_title_dictionary as it is not needed anymore.

**Comment by tonywu1999** [comment]
@datapythonista 

Are there any additional steps I should take to get the pull request approved?  I'm not sure where to go from here.

File: scripts/validate_rst_title_capitalization.py [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+        "-": r"^(?:-{1})*$",
+        "^": r"^(?:\^{1})*$",
+        "~": r"^(?:~{1})*$",
+        "#": r"^(?:#{1})*$",
+        '"': r'^(?:"{1})*$',
+    }
+
+    table = str.maketrans("", "", "*`_")
+
+    for line_no in range(1, len(lines)):
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: I think it's more readable to use `for i, line in enumerate(lines):` and then access the previous line with `lines[i - 1]`.

Using regex also is probably overcomplicating things. You can do something like:
```python
line_chars = set(line)
if len(line_chars) == 1 and line_chats.pop() in ('=', '*',...) and len(line) == len(line[i - 1]):
  ... 
```

File: scripts/validate_rst_title_capitalization.py [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+
+    for directory_address in source_paths:
+        if not os.path.exists(directory_address):
+            raise ValueError(
+                "Please enter a valid path, pointing to a valid file/directory."
+            )
+        elif directory_address.endswith(".rst"):
+            yield directory_address
+        else:
+            for (dirpath, _, filenames) in os.walk(directory_address):
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Can you check if the `glob` module can make this easier?

File: scripts/validate_rst_title_capitalization.py [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+        Heading string to correct
+
+    Returns
+    -------
+    correct_title : str
+        Correctly capitalized heading
+
+    """
+
+    correct_title: str = re.sub(r"^\W*", "", title).capitalize()
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Can you add a comment here, and may be the next two lines, to explain what's going on? It's not immediate to understand what those lines are doing.

File: scripts/validate_rst_title_capitalization.py [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+
+
+def correct_title_capitalization(title: str) -> str:
+    """
+    Algorithm to create the correct capitalization for a given title
+
+    Parameters
+    ----------
+    title : str
+        Heading string to correct
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Can you finish this sentence and the description of returns with a period please.

**Review by datapythonista — COMMENTED** [review]
Thanks, looks great. Just one section feels a bit overcomplicated, but other than that looks really great.

**Comment by tonywu1999** [comment]
@datapythonista

Thank you for your input!  I made the changes that you requested 3 days ago.

File: scripts/validate_rst_title_capitalization.py [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+
+    with open(rst_file, "r") as file_obj:
+        lines = file_obj.read().split("\n")
+
+    symbols = ("*", "=", "-", "^", "~", "#", '"')
+
+    table = str.maketrans("", "", "*`_")
+
+    for i, line in enumerate(lines):
+        if len(line) != 0 and len(lines[i - 1]) != 0:
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: `if len(line) != 0` should be simply `if line` (both cases)

File: scripts/validate_rst_title_capitalization.py [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+            line_chars = set(line)
+            if len(line_chars) == 1 and line_chars.pop() in symbols:
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Why is not as simple as checking that besides this condition the line and the previous have the same length?

I guess I'm missing something, but can't see why all next lines are required, sorry.

**Review by datapythonista — APPROVED** [review]
A question and a small style comment, but lgtm, thanks

File: scripts/validate_rst_title_capitalization.py (hunk 25 above)
Comment (tonywu1999) [reply scripts/validate_rst_title_capitalization.py]: I read the RST documentation guide on what is considered a heading, and there are two situations. 
Case 1:
The heading is between two lines like so:
*******
heading
*******
In this situation, the heading has the be the same length as the length of the line below it.

Case 2:
Heading
***************
In this case, there are no stars above the heading.  In this situation, the heading length only needs to be less than or equal to the length of the line below it to be considered a heading in rST

File: scripts/validate_rst_title_capitalization.py (hunk 24 above)
Comment (tonywu1999) [reply scripts/validate_rst_title_capitalization.py]: @datapythonista made the correction that you requested.  Thank you for your help in reviewing my pull request

File: scripts/validate_rst_title_capitalization.py (hunk 25 above)
Comment (datapythonista) [reply scripts/validate_rst_title_capitalization.py]: I think we always use case 2, with the same length in both lines. I'd personally ignore all options of rst and make the script simpler. What do you think?

Comment (tonywu1999) [reply scripts/validate_rst_title_capitalization.py]: Ok that sounds good to me.  I'll make the changes soon

Comment (tonywu1999) [reply scripts/validate_rst_title_capitalization.py]: @datapythonista made the appropriate changes.

**Comment by tonywu1999** [comment]
Thank you for your guidance @datapythonista

**Comment by datapythonista** [comment]
@pandas-dev/pandas-core CI is red to show how the errors look like, if someone can review we can fix the errors and merge.

**Comment by WillAyd** [comment]
I'm +/- 0 on this. I don't know that the errors highlighted really are all errors, nor that its worth taking this on in pandas. 

But don't oppose strongly if others feel this is moving in the right direction

**Comment by datapythonista** [comment]
@tonywu1999 can you add to the description the issue you're addressing here please? @WillAyd we've got some discussion about this, and one of the main issues is that even in the navigation we have inconsistencies, like `pandas code style` and `Pandas Maintenance` next to each other.

The agreement was to not capitalize more than the first letter, and fix al titles.

**Comment by tonywu1999** [comment]
@datapythonista when you say description, do you mean in the first comment block of my code?  Also do you want me to simply add "GH 26941" to the description?

**Comment by datapythonista** [comment]
I mean in this github issue, not in the code, sorry

**Comment by tonywu1999** [comment]
@datapythonista I changed the description of this issue.  Let me know if you need anything else!

**Comment by datapythonista** [comment]
That is the title. ;)

I updated the description myself. This gives context to reviewers who otherwise think there has been no previous discussion. And also, this will make github close the issue once this is merged.

**Comment by tonywu1999** [comment]
@datapythonista Haha my bad.  Thank you for your help.

**Comment by jbrockmendel** [comment]
@tonywu1999 looks like some (maybe new?) docstrings dont pass this check.  can you update

**Comment by datapythonista** [comment]
@tonywu1999 can you fix the titles you left to see how the errors look like? I think we can get this merged once it's green.

**Comment by datapythonista** [comment]
Also, please merge master into your branch, so we make sure we're validating the latest version of the docs. Thanks!

**Comment by tonywu1999** [comment]
Hi @datapythonista 

I fixed the titles just now and committed the changes.  There should be title changes in doc/source/development/contributing.rst

**Review by datapythonista — APPROVED** [review]
Thanks, can you have a look at the CI? I guess it's unrelated, but if you can merge master and make sure it's green, well be able to merge this.

**Comment by tonywu1999** [comment]
@datapythonista 
I'm not sure what you meant by that.  I merged with master and CI/Checks, which my code affects, is running fine.  Are you referring to the two X's for pandas-dev.pandas and pandas-dev.pandas(Linux py36_locale) ?

**Comment by datapythonista** [comment]
Yes, something failed in that build. Probably something unrelated to this pr, but we need to have it green to merge this. I'm in my phone and can't look at the logs right now, but if you can have a look and see what's the problem, that would be great.

**Comment by tonywu1999** [comment]
@datapythonista 

I looked at the error logs. However, I'm not really sure what's causing the errors.

**Comment by datapythonista** [comment]
> I looked at the error logs. However, I'm not really sure what's causing the errors.

If you navigate all the way to the azure pipeline logs, and look for errors, you'll find this line:
```
AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning('unclosed <ssl.SSLSocket [closed] fd=16, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM,
```

Which looks like a systems error, that (hopefully) shouldn't repeat if the job is executed again. If you merge master into your branch, and push, the CI will run again, and hopefully everything will be green this time. And we'll be able to merge this with confidence.

**Comment by tonywu1999** [comment]
@datapythonista 

I think all checks have passed now.

File: scripts/validate_rst_title_capitalization.py [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+        Heading string to correct.
+
+    Returns
+    -------
+    correct_title : str
+        Correctly capitalized heading.
+
+    """
+
+    # Function to strip all non-word characters from the beginning of the title to the
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Do you mind removing the "Function to" part of all those? I find it a bit confusing, since it's not a function.

Just the "Strip all non-word..." should be enough.

File: scripts/validate_rst_title_capitalization.py [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+
+    line_number : int
+        The corresponding line number of the heading.
+
+    """
+
+    with open(rst_file, "r") as file_obj:
+        lines = file_obj.read().split("\n")
+
+    symbols = ("*", "=", "-", "^", "~", "#", '"')
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Probably worth making this a constant and moving it to the top of the file with the list of words with fiixed capitalization.

File: scripts/validate_rst_title_capitalization.py [hunk 28]
```diff
@@ -0,0 +1,211 @@
...
+    table = str.maketrans("", "", "*`_")
+
+    for i, line in enumerate(lines):
+        line_chars = set(line)
+        if (
+            len(line_chars) == 1
+            and line_chars.pop() in symbols
+            and len(line) == len(lines[i - 1])
+        ):
+            yield lines[i - 1].translate(table), i
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: I'd personally prefer to use:
```python
yield re.sub('[`\*_]', '', lines[i - 1]), i
```
Should be the same, but I think `str.translate` is rarely used and not many people knows about it, and using the regex equivalent should be easier to read for most people.

File: scripts/validate_rst_title_capitalization.py [hunk 29]
```diff
@@ -0,0 +1,211 @@
...
+    title : str
+        A heading found in the rst file.
+
+    line_number : int
+        The corresponding line number of the heading.
+
+    """
+
+    with open(rst_file, "r") as file_obj:
+        lines = file_obj.read().split("\n")
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: I think the standard way to do this would be:
```python
with open(rst_file, "r") as fd:
    for i, line in enumerate(fd):
     ....
```

File: scripts/validate_rst_title_capitalization.py [hunk 30]
```diff
@@ -0,0 +1,211 @@
...
+    yields the strings of the .rst file directories that these paths contain.
+
+    Parameters
+    ----------
+    source_paths : str
+        List of directories to validate, provided through command line arguments.
+
+    Yields
+    -------
+    directory_address : str
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]:
```suggestion
    str
```

File: scripts/validate_rst_title_capitalization.py [hunk 31]
```diff
@@ -0,0 +1,211 @@
...
+        Directory address of a .rst files found in command line argument directories.
+
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Better remove this blank line (we're not validating this docstring, but this blank line would make it fail if we do.

File: scripts/validate_rst_title_capitalization.py [hunk 32]
```diff
@@ -0,0 +1,211 @@
...
+    Algorithm to create the correct capitalization for a given title.
+
+    Parameters
+    ----------
+    title : str
+        Heading string to correct.
+
+    Returns
+    -------
+    correct_title : str
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]:
```suggestion
    str
```

File: scripts/validate_rst_title_capitalization.py [hunk 33]
```diff
@@ -0,0 +1,211 @@
...
+    "Parquet",
+    "MultiIndex",
+    "NumFOCUS",
+    "sklearn",
+    "Docker",
+}
+
+CAP_EXCEPTIONS_DICT = {word.lower(): word for word in CAPITALIZATION_EXCEPTIONS}
+
+err_msg = "Heading capitalization formatted incorrectly. Please correctly capitalize"
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Since we're using this just once, I'd move it to the function where it's being used.

File: scripts/validate_rst_title_capitalization.py [hunk 34]
```diff
@@ -0,0 +1,211 @@
...
+    Parameters
+    ----------
+    rst_file : str
+        RST file to scan through for headings.
+
+    Yields
+    -------
+    title : str
+        A heading found in the rst file.
+
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: No blank line here

File: scripts/validate_rst_title_capitalization.py [hunk 35]
```diff
@@ -0,0 +1,211 @@
...
+    line_number : int
+        The corresponding line number of the heading.
+
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: No blank line here either.

File: scripts/validate_rst_title_capitalization.py [hunk 36]
```diff
@@ -0,0 +1,211 @@
...
+    source_paths : str
+        List of directories to validate, provided through command line arguments.
+
+    Yields
+    -------
+    directory_address : str
+        Directory address of a .rst files found in command line argument directories.
+
+    """
+
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: No blank line.

File: scripts/validate_rst_title_capitalization.py [hunk 37]
```diff
@@ -0,0 +1,211 @@
...
+    output_format : str
+        Output format of the script.
+
+    Returns
+    -------
+    number_of_errors : int
+        Number of incorrect headings found overall.
+
+    """
+
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: No blank line

File: scripts/validate_rst_title_capitalization.py [hunk 38]
```diff
@@ -0,0 +1,211 @@
...
+    Parameters
+    ----------
+    source_paths : str
+        List of directories to validate, provided through command line arguments.
+    output_format : str
+        Output format of the script.
+
+    Returns
+    -------
+    number_of_errors : int
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]:
```suggestion
    int
```

File: scripts/validate_rst_title_capitalization.py [hunk 39]
```diff
@@ -0,0 +1,211 @@
...
+        Number of incorrect headings found overall.
+
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: No blank line

**Review by datapythonista — COMMENTED** [review]
Thanks @tonywu1999, and sorry for the delay.

I added few minor final comments more if you don't mind, I think they'll improve readability and future maintenance.

File: scripts/validate_rst_title_capitalization.py [hunk 40]
```diff
@@ -0,0 +1,206 @@
...
+    -------
+    str
+        Correctly capitalized heading.
+    """
+
+    # Strip all non-word characters from the beginning of the title to the
+    # first word character.
+    correct_title: str = re.sub(r"^\W*", "", title).capitalize()
+
+    # Function to remove a URL from the title. We do this because words in a URL must
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: Can you also remove the "Function to" from here.

File: scripts/validate_rst_title_capitalization.py [hunk 41]
```diff
@@ -0,0 +1,206 @@
...
+    # stay lowercase, even if they are a capitalization exception.
+    removed_https_title = re.sub(r"<https?:\/\/.*[\r\n]*>", "", correct_title)
+
+    # Function to split a title into a list using non-word character delimiters.
```
Comment (datapythonista) [inline scripts/validate_rst_title_capitalization.py]: And from here.

Sorry, I thought I added it to the comment in the first occurrence.

**Review by datapythonista — COMMENTED** [review]
I see there are still some pending comments from the previous review, let me know when this is ready to be reviewed again.

**Comment by datapythonista** [comment]
Thanks for this @tonywu1999, good improvement.

Can you please open an issue to fix the titles in the rest of the files?

**Comment by tonywu1999** [comment]
@datapythonista 

Thanks for your guidance! I opened a new issue (#32550) to address fixing the titles in the rest of the files
